tracer/build/_build/Build.VariableGenerations.cs

```diff
@@ -0,0 +1,263 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using Nuke.Common;
+using Nuke.Common.CI.AzurePipelines;
+using Nuke.Common.Execution;
+using Nuke.Common.Tools.Git;
+using NukeExtensions;
+
+partial class Build : NukeBuild
+{
+
+    [Parameter("Indicates matrices target. Match the name of generated variable", List = false)]
+    readonly GenerateMatricesTarget? GenerateMatricesTarget;
+
+    [Parameter("Indicates condition. Must be in the format '{variableName}|[filter1,filter2];{variableName}|[filter1,filter2]'", List = false)]
+    readonly string GenerateConditionVariableFilter;
+
+    Target GenerateMatrices
+        => _ =>
+        {
+            return _
+                  .Unlisted()
+                  .Requires(() => GenerateMatricesTarget)
+                  .Executes(() =>
+                  {
+                      if (!CheckChanges())
+                      {
+                          Logger.Info($"No changes found.");
+                          return;
+                      }
+
+                      var target = GenerateMatricesTarget ?? global::GenerateMatricesTarget.none;
+                      if (global::GenerateMatricesTarget.integration_tests_windows_matrices.HasFlag(target))
+                      {
+                          GenerateIntegrationTestsWindowsMatrices(target);
+                      }
+
+                      if (target.HasFlag(global::GenerateMatricesTarget.integration_tests_linux_matrix))
+                      {
+                          GenerateIntegrationTestsLinuxMatrix();
+                      }
+
+                      if (global::GenerateMatricesTarget.exploration_tests_matrices.HasFlag(target))
+                      {
+                          GenerateExplorationTestMatrices(target);
+                      }
```

Comment:
I think the attempt to generalise this code is making it much more complex than it needs to be. The pipeline is always going to be tightly coupled to this method, so I would just hard code everything in here. Always generate all the matrices, and move all the logic into Nuke (there shouldn't be be much left after you do that). I think this first iteration is much harder to grok than the bash currently, though it's definitely a step in the right direction!

Reply:
👍  From my perspective, there is no rush to merge this into `master`, so it makes sense to iterate on it in this PR.

Reply:
I agree, that it looks a little bit complex at first view.
However, I want to achieve following:

1. Visibility for yml. In yml, we are required to use in the way:
`matrix: $[ stageDependencies.generate_matrices.generate_matrices_job.outputs['generate_matrices_step.exploration_tests_linux_matrix'] ]`

All parts of conditions are clear, and can be found in the same file, except of "exploration_tests_linux_matrix" variable name. If remove this parameter from yml file, it will be totally unclear, where it came from.
Same for variables and conditions, like `isDebuggerChanged`

2. Testability. The way it implemented now, you can easily run nuke with any custom condition, by simply changing input for nuke parameter, without changing any line of code. 

I think those are important things.
I can reduce complexity, by removing flags from enum.

Reply:
> 1. Visibility for yml... If remove this parameter from yml file, it will be totally unclear

I strongly disagree 🙂 It's obvious it comes from Nuke, so doesn't seem like a big step to check there? And the massive coupling adds a lot of unnecessary coupling and complexity IMO

> 2. Testability. The way it implemented now, you can easily run nuke with any custom condition

I don't follow this one, do you mean testing the generator? If so, always generating all of the matricies doesn't seem to impact testability?

Reply:
Ok, I've got it.
Removed.